### PR TITLE
Add grant/revoke for user access to a model.

### PR DESCRIPTION
**Pull request by kadams54**

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6766/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6767/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6768/

**Comment by kadams54** [comment]
No QA, as these functions aren't yet hooked into a UI. Just need reviews.

File: jujugui/static/gui/src/app/store/env/controller-api.js [hunk 1]
```diff
@@ -1261,6 +1261,96 @@ YUI.add('juju-controller-api', function(Y) {
         request: 'RevokeCredentials',
         params: {entities: [{tag: tags.build(tags.CREDENTIAL, name)}]}
       }, handler);
+    },
+
+    /**
+      Modify (grant or revoke) user access to the specified model.
+
+      @method grantUserAccess
```
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/controller-api.js]: s/grantUserAccess/_modifyModelAccess/

File: jujugui/static/gui/src/app/store/env/controller-api.js [hunk 2]
```diff
@@ -1261,6 +1261,96 @@ YUI.add('juju-controller-api', function(Y) {
...
+      @param {Array} userIds The IDs of users who need modified access.
```
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/controller-api.js]: `@param {Array} users The usernames of the users who need modified access.`

File: jujugui/static/gui/src/test/test_controller_api.js [hunk 3]
```diff
@@ -2077,4 +2077,181 @@ describe('Controller API', function() {
     });
   });
 
+  describe('grantModelAccess', function() {
+    it('grants users access', function(done) {
+      // Perform the request.
+      const users = ['dalek', 'drwho@external'];
+      const access = 'read';
+      controllerAPI.grantModelAccess(users, 'default', access, err => {
```
Comment (frankban) [inline jujugui/static/gui/src/test/test_controller_api.js]: 'default' here is kind of misleading as it should be a model uuid. Maybe use 'uuid-1234' or similar.

File: jujugui/static/gui/src/app/store/env/controller-api.js [hunk 4]
```diff
@@ -1261,6 +1261,96 @@ YUI.add('juju-controller-api', function(Y) {
...
+          return;
+        }
+        if (data.error) {
+          callback(data.error);
+          return;
+        }
+        const results = data.response.results;
+        if (!results || results.length !== 1) {
+          // This should never happen.
+          callback('Invalid results from Juju: ' + JSON.stringify(results));
```
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/controller-api.js]: s/Invalid/invalid/ to mimic Juju error style, which allows for concatenating error messages.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6774/

**Comment by kadams54** [comment]
Thanks @frankban and @jaycee. :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**